### ISSUE TITLE
DOI Support - Ability to assign DOIs to datasets

### ISSUE THREAD

**Issue by rufuspollock** [ckan/ideas#6]
Support for assigning DOIs to datasets in a CKAN instance and displaying this appropriately to encourage use in citation.
- DataCite API is at: https://mds.datacite.org/static/apidoc
- EZID (DOI as well) API: http://n2t.net/ezid/doc/apidoc.html
## Colophon

Original item (with lots of discussion) at: https://trello.com/c/dlzGQGdH/98-discovery-ability-to-assign-dois-datacite-to-datasets

**Comment by makkus** [comment]
Is anybody currently working on this? I've seen the ckanext-doi project created by NHM, but that seems to only support the Datacite API. We are using EZID. Might not be too hard to add EZID support for that, so I'll give it a go. Just wondering whether somebody already tried...

**Comment by florianm** [comment]
I've spoken to the Australian National Data Service (ANDS) who are willing to offer [DOI minting](http://ands.org.au/services/cite-my-data.html). 
I can't promise it'll be free of charge but for our use case (WA govt [data](http://catalogue.beta.data.wa.gov.au/)) it might be.

The discussion point re ANDS would be policy of minting DOIs (per dataset, per resource, per resource revision?) and cost if any.

The remaining task is to develop ckanext-doi from using datacite into using ANDS as well (service and credentials via config). I've got that on my wish list as soon as I've got playtime left; happy to collaborate and communicate though.

Getting a citation metric back shown on datasets with a DOI would be a nice to have. Surely there's an API we could ping from the package_read template.

**Comment by Aaron-M** [comment]
We are working with the CKAN developers (contact Jo Barratt) at present on a DOI module that works with the EZID service (which we are signed up to issue DOIs through). We are setting it up and wil test it/install on our system in due course, but the idea would be that other users would need to have their own EZID account in order to issue their own DOIs.  ANDS providing that service for the whole of Australian Research Orgs is a slightly different use case, but the underlying functionaility we are working on could be shared?

**Comment by makkus** [comment]
Yes, that'd be good. I was thinking along the lines of having a generic '(persistent)Id' base-plugin, which could be extended to connect to different 'pid' providers (EZID, datacite, a random handle server, maybe even a 'normal' url or bit.ly or similar -- even though of course a 'normal' id would not be really persistent, but it could be useful anyway).

**Comment by grhammo** [comment]
Hi all, any updates on this one? 
My organisation would really like to provide a DOI for each dataset, so that we can ensure research citation is represented accurately.

**Comment by makkus** [comment]
My colleague @neon-ninja is now working on a plugin to add EZID DOIs to CKAN resources:

https://github.com/UoA-eResearch/ckanext-ezid

Actually, it's more than one plugin. There is a base one ( https://github.com/UoA-eResearch/ckanext-external_id ) that allows for generic ids to be added to a resource, and which needs to be extended by id-type-specific plugins like the one above. There is also an example implementation of such a plugin that can do url-shortening via goo.gl: https://github.com/UoA-eResearch/ckanext-googl

Feel free to contribute, or develop your own doi-provider specific plugin (if you are not using EZID).

**Comment by brew** [comment]
Great to see some work on this. We (Open Knowledge) have also been working on an EZID based DOI extension for datasets: https://github.com/okfn/ckanext-doi .

It provides an additional form field for automatically generating an DOI for a dataset and submitting to EZID when the dataset is published. Or the DOI can be manually inputted. There's a setting to define which roles within an organization can publish a DOI, so for example it can be restricted to org admins only. There's also support for selecting from a list of predefined prefixes (shoulders).

**Comment by Psykar** [comment]
For those in Australia, I've just uploaded an extension to work with ANDS for minting DOIs

pypi.python.org/pypi/ckanext-ands
https://github.com/Psykar/ckanext-ands
